x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/movingPercentiles/MovingPercentilesPipelineAggregationBuilder.java

```diff
@@ -0,0 +1,129 @@
+import org.elasticsearch.search.aggregations.pipeline.AbstractPipelineAggregationBuilder;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+
+public class MovingPercentilesPipelineAggregationBuilder
+        extends AbstractPipelineAggregationBuilder<MovingPercentilesPipelineAggregationBuilder> {
+    public static final String NAME = "moving_percentiles";
+    private static final ParseField WINDOW = new ParseField("window");
+    private static final ParseField SHIFT = new ParseField("shift");
+
+    public static final ConstructingObjectParser<MovingPercentilesPipelineAggregationBuilder, String> PARSER =
+            new ConstructingObjectParser<>(NAME, false, (args, name) -> {
+                return new MovingPercentilesPipelineAggregationBuilder(name, (String) args[0], (int) args[1]);
+            });
+    static {
+        PARSER.declareString(constructorArg(), BUCKETS_PATH_FIELD);
+        PARSER.declareInt(constructorArg(), WINDOW);
+        PARSER.declareInt(MovingPercentilesPipelineAggregationBuilder::setShift, SHIFT);
+    }
+
+    private final int window;
+    private int shift;
+
+    public MovingPercentilesPipelineAggregationBuilder(String name, String bucketsPath, int window) {
+        super(name, NAME, new String[] { bucketsPath });
+        if (window <= 0) {
+            throw new IllegalArgumentException("[" + WINDOW.getPreferredName() + "] must be a positive, non-zero integer.");
+        }
+        this.window = window;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public MovingPercentilesPipelineAggregationBuilder(StreamInput in) throws IOException {
+        super(in, NAME);
+        window = in.readVInt();
+        shift = in.readInt();
+    }
+
+    @Override
+    protected final void doWriteTo(StreamOutput out) throws IOException {
+        out.writeVInt(window);
+        out.writeInt(shift);
+    }
+
+    /**
+     * Returns the window size for this aggregation
+     */
+    public int getWindow() {
+        return window;
+    }
+
+    /**
+     * Returns the shift for this aggregation
+     */
+    public int getShift() {
+        return shift;
+    }
+
+    /**
+     * Sets the shift for this aggregation
+     */
+    public void setShift(int shift) {
+        this.shift = shift;
+    }
+
+    @Override
+    protected PipelineAggregator createInternal(Map<String, Object> metaData) {
+        return new MovingPercentilesPipelineAggregator(name, bucketsPaths, getWindow(), getShift(), metaData);
+    }
+
+    @Override
+    protected void validate(ValidationContext context) {
+        if (bucketsPaths.length != 1) {
+            context.addBucketPathValidationError("must contain a single entry for aggregation [" + name + "]");
+        }
+        context.validateParentAggSequentiallyOrdered(NAME, name);
```

Comment:
I think it may be worth adding something near `bucketCardinality` that returns the type of buckets. That'd let us do this `ordered` assertion without `instanceof`s. I'd be happy to take a stab at it in a follow up because I have ideas. No need to delay this PR for it though.